bitmark-references/bitmark-node-software/mining.md

```diff
@@ -0,0 +1,123 @@
+
+When `bitmarkd` receives this message from `recorderd`, the `bitmarkd` validates the hash from this special message and returns the result to the `recorderd`.
+
+If `bitmarkd` receives no valid hashes from other nodes or from `recorderd`, then the `bitmarkd` will periodically send hash tasks to `recorderd` (currently every 1 minute), until a valid hash is found or received.
+
+## Hashing
+
+Hashing is the transformation of a string of characters into an unique value or key that represents the original string. There are different hash implementations with a wide range of algorithm, if the hash implementation is well designed, it turns to be very difficult to guess what is the original data from a given hash value.
+
+At bitmark we use `argon2` to compute the hashing, the `argon2` has some benefits such as a memory hard algorithm that provides resistance against GPU and ASIC hardware for computation as well as it provides a hash of 256 bits long.
+
+The `argon2` user-land software provides a wide range of parameters, here there is an example:
+
+```
+printf '%s' 'hello world' | argon2 'hello world' -d -l 32 -m 17 -t 4 -p 1 -r | awk '{for(i=length($1);i>0;i-=2)x=x substr($1,i-1,2);print x}'
+
+f8a17bc25cb53e848e2d09811ade4b8a037f628443661b88611faf5d9a5a1f33
+```
+
+A block hash contains the following data:
+
+1. block record version
+
+    blockchain header version
+
+1. transaction count
+
+    number of transactions in the block
+
+1. merkle tree
+
+    there is an allowable limit of maximum 9999 transactions hashes per block, if the number of verified transactions are above the limit, random transactions are selected.
+
+1. timestamp
+
+    block generated time
+
+1. difficulty
+
+    hashing difficulty
+
+1. nonce
+
+    random string to make the hash fits the difficuty
+
+## Difficulty
+
+It is possible to set the level of difficulty to find a valid hash, higher difficulty means harder to find a valid hash, when a hash meets difficulty, it means the hash value is less than or equal to the difficulty level. Usually the difficulty level is represented by a number.
+
+The hash is considered as a fixed 256 bits value and it is composed by two parts:
+
+    [8 bits] [248 bits]
+    exponent mantissa
+
+Difficulty value is encoded by 64 bits, including 8 bits of exponent and 56 bits of mantissa.
+
+    [56 bits] [8 bits]
+    mantissa  exponent = 64 bits unsigned value
+
+The mantissa bits are normalized, the most significant bit is 1 and can be dropped, leaving the 56 bits to store; these bits are right shifted by exponent +8 from the most significant bit.
+
+The difficulty `One` is defined as `00 ff  ff ff  ff ff  ff ff`, represents 256 bits value `0ff ffff ffff ffff 8000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000`.
+
+The `Two` is the double of difficulty `One`: `01ffffffffffffff`, represents 256 bits value `007fffffffffffffc00000000000000000000000000000000000000000000000`.
+
```

Comment:
I feel like this section is missing two things:

1) It just says "It is possible to set the level of difficulty". Who sets this difficulty and why? (I assume it's set by the blockchain consensus rules, and is issued by bitmarkd as part of the hashing job? Whatever it is, we should say.)

2.) It's similarly brief on how the difficulty actually impacts the hashing. I know that "when a hash meets difficulty, it means the hash value is less than or equal to the difficulty level."  explains it, but we really need an example to make things clear.

Reply:
updated

Reply:
Example is great.